org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/diagnostics/LSPDiagnosticsToMarkers.java

```diff
@@ -194,4 +194,42 @@ private IMarker getExistingMarkerFor(IDocument document, Diagnostic diagnostic,
 		}
 		return null;
 	}
+
+	private @NonNull Map<String, Object> computeMarkerAttributes(@Nullable IDocument document,
+			@NonNull Diagnostic diagnostic, @NonNull IResource resource) {
+		Map<String, Object> attributes = new HashMap<>(8);
+		attributes.put(LSP_DIAGNOSTIC, diagnostic);
+		attributes.put(LANGUAGE_SERVER_ID, languageServerId);
+		attributes.put(IMarker.MESSAGE, diagnostic.getMessage());
+		attributes.put(IMarker.SEVERITY, LSPEclipseUtils.toEclipseMarkerSeverity(diagnostic.getSeverity()));
+
+		if (document != null) {
+			Range range = diagnostic.getRange();
+			int documentLength = document.getLength();
+			try {
+				int start = Math.min(LSPEclipseUtils.toOffset(range.getStart(), document), documentLength);
+				int end = Math.min(LSPEclipseUtils.toOffset(range.getEnd(), document), documentLength);
+				int lineOfStartOffset = document.getLineOfOffset(start);
+				if (start == end && documentLength > end) {
+					end++;
+					if (document.getLineOfOffset(end) != lineOfStartOffset) {
+						start--;
+						end--;
+					}
+				}
+				attributes.put(IMarker.CHAR_START, start);
+				attributes.put(IMarker.CHAR_END, end);
+				attributes.put(IMarker.LINE_NUMBER, lineOfStartOffset + 1);
+			} catch (BadLocationException ex) {
+				LanguageServerPlugin.logError(ex);
+			}
+		}
+
+
+		if (markerAttributeComputer != null) {
+			markerAttributeComputer.addMarkerAttributesForDiagnostic(diagnostic, document, resource, attributes);
```

Comment:
my JDT null analysis tells me there is a potential NPE in here, it looks like a bug in the null analysis of eclipse because we check for null just before. The marker is annoying and it will be annoying for anyone fetching the codebase, so I have changed it to use Optional.